enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,278 @@
+encrypted with those keys--but users would expect the persistent storage volumes to
+be encrypted as well. This enhancement proposes a solution for AWS, GCP, & Azure.
+
+## Motivation
+
+The primary motivation is to enable managed services to encrypt storage volumes with
+customer-provided keys automatically as part of the installation process without requiring
+day-2 operations. The behavior of automatically setting the default storage class is
+sensible for individual (non-managed-service) users as well; setting the default will
+reduce their day-2 burden and decrease the opportunity for user error.
+
+### User Stories
+
+When I install a cluster, I want to provide user-managed encryption keys in the install-config
+so that the default storage class will be set to provision volumes with those keys.
+
+### Goals
+
+* Default storage class is set to use user-provided keys at cluster creation
+
+### Non-Goals
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer adds encryption key fields to cluster infrastructure object.
+3. Cluster-storage-operator/CSI drivers use cluster infrastructure object
+to create default storage class using user-encrypted keys from infra object.
+
+
+### API Extensions
+
+This enhancement extends the cluster `Infrastructure` API object to add fields
+for the user-managed encryption keys. The fields will be added to the platform
+spec and status for each platform. They are added to the spec so that users can edit the
+spec day 2 in order to update the default storage class. The status field would
+be updated by the cluster-storage operator/CSI driver.
+
+#### AWS 
+
+```go
+// AWSPlatformSpec holds the desired state of the Amazon Web Services infrastructure provider.
+// This only includes fields that can be modified in the cluster.
+type AWSPlatformSpec struct {
+	// [... existing fields ...]
+
+    // The KMS key to be used in the default storage class to encrypt volumes.
+    // +optional
+    KMSKeyARN string `json:"kmsKeyARN,omitempty"`
+}
+
+// AWSPlatformStatus holds the current status of the Amazon Web Services infrastructure provider.
+type AWSPlatformStatus struct {
+    // [... existing fields ...]
+
+    // The KMS key used by the default storage class to encrypt volumes.
+    // +optional
+    KMSKeyARN string `json:"kmsKeyARN,omitempty"`
+}
+```
+
+#### Azure
+
+```go
+// AzureDiskEncryptionSet defines the configuration for a disk encryption set.
+type AzureDiskEncryptionSet struct {
+  // SubscriptionID defines the Azure subscription that contains the disk encryption set.
+  SubscriptionID string `json:"subscriptionId"`
+  
+  // ResourceGroup defines the Azure resource group that contains the disk encryption set.
+  ResourceGroup string `json:"resourceGroup"`
+  
+  // Name is the name of the disk encryption set.
+  Name string `json:"name"`
+}
```

Comment:
Azure Disk CSI driver supports the key in this format, which is somehow comparable to an ARN in AWS: `/subscriptions/{subs-id}/resourceGroups/{rg-name}/providers/Microsoft.Compute/diskEncryptionSets/{diskEncryptionSet-name}`.

We can easily build this string with the information provided in this API, however, would it make sense to have a single field with the required format instead (analogous to the ARN field in AWS)?

Also, the CSI driver supports the field `diskEncryptionType`, [which specifies the encryption type](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/driver-parameters.md) of the encryption set. It seems like this could be also useful here?

Reply:
> We can easily build this string with the information provided in this API, however, would it make sense to have a single field with the required format instead (analogous to the ARN field in AWS)?

Personally I would stick with the encryption set API key as currently documented, but I do not feel strongly about this. 

> Also, the CSI driver supports the field diskEncryptionType, [which specifies the encryption type](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/driver-parameters.md) of the encryption set. It seems like this could be also useful here?

Perhaps we should just stick with the default which would be `EncryptionAtRestWithCustomerKey`, then we would not need to include this in the API. If users desired the other option, `EncryptionAtRestWithPlatformAndCustomerKeys`, can they change the default storage class?

Reply:
Is it not a better UX to have the components broken down and then compiled into the ARN like string by the system? If we were to change either Azure or AWS, I would probably break AWS down like the Azure one is